services/sharedchannel/sync_send.go

```diff
@@ -0,0 +1,260 @@
+	for {
+		select {
+		case <-scs.changeSignal:
+			for {
+				task, ok := scs.removeOldestTask()
+				if !ok {
+					break
+				}
+				if err := scs.processTask(task); err != nil {
+					scs.server.GetLogger().Error("Failed to update shared channel", mlog.String("channelId", task.channelId), mlog.String("remoteId", task.remoteId), mlog.Err(err))
+					// put task back into map so it will update again
+					scs.addTask(task)
+				}
+			}
+		case <-done:
+			return
+		}
+	}
+}
+
+// removeOldestTask removes and returns the oldest task in the task map.
+// If no tasks are available then false is returned.
+func (scs *Service) removeOldestTask() (syncTask, bool) {
+	scs.mux.Lock()
+	defer scs.mux.Unlock()
+
+	var oldestTask syncTask
+	var oldestKey string
+
+	for k, v := range scs.tasks {
+		if v.AddedAt.Before(oldestTask.AddedAt) || oldestTask.AddedAt.IsZero() {
+			oldestKey = k
+			oldestTask = v
+		}
+	}
+
+	if oldestKey != "" {
+		delete(scs.tasks, oldestKey)
+		return oldestTask, true
+	}
+	return oldestTask, false
+}
+
+// processTask updates one or more remote clusters with any new channel content.
+func (scs *Service) processTask(task syncTask) error {
+	var err error
+	var remotes []*model.RemoteCluster
+
+	if task.remoteId == "" {
+		filter := model.RemoteClusterQueryFilter{
+			InChannel:     task.channelId,
+			OnlyConfirmed: true,
+		}
+		remotes, err = scs.server.GetStore().RemoteCluster().GetAll(filter)
+		if err != nil {
+			return err
+		}
+	} else {
+		rc, err := scs.server.GetStore().RemoteCluster().Get(task.remoteId)
+		if err != nil {
+			return err
+		}
+		if !rc.IsOnline() {
+			return fmt.Errorf("Failed updating shared channel '%s' for offline remote cluster '%s'", task.channelId, rc.DisplayName)
+		}
+		remotes = []*model.RemoteCluster{rc}
+	}
+
+	cache := make(msgCache)
+	errs := merror.New()
+	for _, rc := range remotes {
+		if err := scs.updateForRemote(task.channelId, rc, cache); err != nil {
+			errs.Append(err)
+			// retry...
+			if task.retryCount < MaxRetries {
+				retryTask := newSyncTask(task.channelId, rc.RemoteId)
+				retryTask.retryCount = task.retryCount + 1
+				scs.addTask(retryTask)
+			}
+		}
+	}
+	return errs.ErrorOrNil()
+}
+
+// updateForRemote updates a remote cluster with any new posts/reactions for a specific
+// channel. If many changes are found, only the oldest X changes are sent and the channel
+// is re-added to the task map. This ensures no channels are starved for updates even if some
+// channels are very active.
+func (scs *Service) updateForRemote(channelId string, rc *model.RemoteCluster, cache msgCache) error {
+	scr, err := scs.server.GetStore().SharedChannel().GetRemoteByIds(channelId, rc.RemoteId)
+	if err != nil {
+		return err
+	}
+
+	opts := model.GetPostsSinceOptions{
+		ChannelId: channelId,
+		Time:      scr.LastSyncAt,
+	}
+	posts, err := scs.server.GetStore().Post().GetPostsSince(opts, true)
+	if err != nil {
+		return err
+	}
```

Comment:
Wouldn't it create less inconsistencies and also perform better if we got the posts just once outside the loop instead of doing it for every remote?

Reply:
The remotes may be wildly different in which posts they are getting per batch.  I did create a caching mechanism so posts don't get serialized repeatedly.  So if two remotes are in fact getting the same set, the post is serialized once, including the reactions.

Reply:
Oh I see, we are leveraging `LastSyncAt`. So what would happen if either some posts get updated while this is running and or reactions are added for already cached posts? Could we have stale reads from cache issues?

Reply:
If a reaction happens while in the middle of a sync, a new sync task would get enqueued for that new reaction.  Each sync task gets a new cache, and it only ensures remotes getting the same update serialize the data once per post for that sync task.

Reply:
Alright so what we've temporarily missed in between would need some different synchronization logic.

Reply:
Discussed this internally. My confusion was around `LastSyncAt` value which is not a timestamp generated by the clusters during the sync process but the value of `UpdateAt` from the latest (newest) synched post. This guarantees we are not going to miss posts in between syncs.